### ISSUE TITLE
Stop hardcoding the offered language choices

### ISSUE THREAD

**Issue by martinpitt** [cockpit-project/cockpit#8439]
Spot when reviewing #8346: We currently hardcode the list of available languages; current master has it in pkg/shell/manifest.json.in, PR #8346  moves it to src/ws/login.html.

This should be generated more dynamically; e. g. at build time it could check which of po/*.po are more than, say, 70% translated, and generate the choices there.

**Comment by petervo** [comment]
IMO we should only add never remove.

**Comment by marusak** [comment]
I was looking into this, but the thing I am stuck at is how to get native language names.

First idea was to get it from Zanata automatically. Even though there is API, I could not make it work in any way to get native version on languages.
Another idea was to get if from `locale`. Problem is between mapping of Zanata language names and locale ones. For example zanata reference `nds` what locale has as `nds_DE`. This can be simply solved, main problem is with languages that cannot be grab from locales, such as zanatas `bal` language.
Last idea I had, was to manually download this list from Zanata and use it directly (for example create python map and put it directly in bots/po-refresh)

Any other ideas?

**Comment by martinpitt** [comment]
@marusak: Language codes and names are defined in [ISO 639](https://en.wikipedia.org/wiki/ISO_639). Fedora (and also Debian) have an "iso-codes" package which ships JSON and XML formats of these. There are also various python modules ([iso639-python](https://github.com/janpipek/iso639-python) and [iso639](https://github.com/noumar/iso639/), both are packaged in Fedora) which makes working with these very easy.

I suggest doing this mapping at build time. This can check which translations are available, build the native language list, and include it into the webpack.

**Comment by marusak** [comment]
So after trying a few libraries, only really usable one is [langcodes](https://github.com/LuminosoInsight/langcodes). Other cannot parse for example `pt_BR`. This library is still not awesome with it because `pt` is `português` and `pt_BR` is still português (it is Brazilian dialect) and therefore not distinguishable.

Next step is to find out how to get from `da` to linux locale `da_DK`. Possibly we can use locales like `danish` (english names)? Or even better - if the lang code from zanata exists in locales then use that otherwise english name. (So `pt_BR` would use locale `pt_BR` and `da` would use `danish`.) And just to be on a safe side, each new language would open separate commit (or even PR) so it can be manually checked and possibly edited.

**Comment by martinpitt** [comment]
@marusak: In my former life I worked a lot on Ubuntu's [language-selector](https://git.launchpad.net/ubuntu/+source/language-selector/tree/), which had to deal with pretty well the same issues. For example, it has a [langcode → locales map](https://git.launchpad.net/ubuntu/+source/language-selector/tree/data/langcode2locale) and a [language list](https://git.launchpad.net/ubuntu/+source/language-selector/tree/data/languagelistwhich) (mapping langcodes to lang name and primary locale) that you could just re-use - the latter is probably more useful.

To summarize from IRC, the large scheme of that would be that we want to do this on bots/po-refresh time. This can slurp in the languages with ≥ 50%  coverage, ignore the others, and build some .js code from a template that gets included into the language selection dialog. Then this can be tested and reviewed, and it happens at the right time.
